tsl/src/remote/dist_commands.c

```diff
@@ -76,6 +78,117 @@ ts_dist_cmd_collect_responses(List *requests)
 	return results;
 }
 
+/*
+ * Invoke SQL statement on the given data nodes without throwing.
+ *
+ * Since this function clones results, you must call ts_dist_cmd_close_response
+ * on its DistCmdResult.
+ *
+ * Todo: Implement search path support
+ */
+DistCmdResult *
+ts_dist_cmd_invoke_on_data_nodes_no_throw(const char *sql, List *data_nodes)
+{
+	ListCell *lc_data_node, *lc_conn, *lc;
+	List *requests = NIL, *connections = NIL;
+	DistCmdResult *results;
+	AsyncRequestSet *rs;
+	AsyncResponseResult *ar;
+
+	if (data_nodes == NIL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("no data nodes to execute command on"),
+				 errhint("Add data nodes before executing a distributed command.")));
+
+	switch (nodeTag(data_nodes))
+	{
+		case T_OidList:
+			data_nodes = data_node_oids_to_node_name_list(data_nodes, ACL_NO_CHECK);
+			break;
+		case T_List:
+			/* Already in the format we want */
+			data_node_name_list_check_acl(data_nodes, ACL_NO_CHECK);
+			break;
+		default:
+			elog(ERROR, "invalid list type %u", nodeTag(data_nodes));
+			break;
+	}
+
+	int i = 0;
+	results = palloc0(sizeof(DistCmdResult) + list_length(data_nodes) * sizeof(DistCmdResponse));
+
+	foreach (lc_data_node, data_nodes)
+	{
+		const char *node_name = lfirst(lc_data_node);
+		char *errmsg;
+		AsyncRequest *req;
+		TSConnection *connection = data_node_get_connection_nothrow(node_name, &errmsg);
+
+		/* Todo: There is no function to set errorMessage in a pg_result.
+		 * For now, copy errorMesage around in DistCmdResponse */
+		if (connection == NULL)
+		{
+			DistCmdResponse *response = &results->responses[i];
+			response->pg_result = PQmakeEmptyPGresult(NULL, PGRES_FATAL_ERROR);
+			response->data_node = pstrdup(node_name);
+			response->errorMessage = pstrdup(errmsg);
+			++i;
+			continue;
+		}
+
+		connections = lappend(connections, connection);
+
+		ereport(DEBUG2, (errmsg_internal("sending \"%s\" to data node \"%s\"", sql, node_name)));
+
+		req = async_request_send(connection, sql);
+
+		async_request_attach_user_data(req, (char *) node_name);
+		requests = lappend(requests, req);
+	}
+
+	rs = async_request_set_create();
+
+	foreach (lc, requests)
+		async_request_set_add(rs, lfirst(lc));
+
+	while ((ar = async_request_set_wait_ok_result(rs)))
+	{
+		DistCmdResponse *response = &results->responses[i];
+
+		/* Closing remote connections will trigger event processing
+		 * which in turn clears PGresults from Async Responses.
+		 * Make a deep copy before closing.
+		 * Todo: a deep copy will always be PGRES_TUPLES_OK */
+		response->pg_result =
+			PQcopyResult(async_response_result_get_pg_result(ar), PG_COPYRES_TUPLES);
```

Comment:
Hmm, I am hesitant about this approach. Ideally, we should life-cycle manage also Result copies, although we aren't at this time because we haven't used any Result copies so far.

In other words, we should probably store the Result copy on the connection list when `PGEVT_RESULTCOPY` occurs, although we currently don't handle that event because of the above reason. However, if we do in the future, then this will break.

The reason we life-cycle manage the Results is because they are malloc'd  instead of allocated on PG memory contexts. So, there's a huge risk of memory leaks otherwise.

Reply:
Maybe one way to fix this is to change the API for this function to require the caller to pass in the connection. Then the connection can be managed outside the scope of the function and one can ensure it is not closed until responses are processed.

Reply:
I thought about that too, will do. There's no harm to keeping the connection in the DistCmdResponse.